src/Store.elm

```diff
@@ -35,21 +34,103 @@ type alias Store =
 
 {-| As in, Request
 -}
-type Action
-    = GetPosts
-    | GetUsers
-    | GetImage ImageId
-    | CreatePost PostCreateData
+type alias Action =
+    { run : Store -> ( Store, Cmd Msg )
+    , toastOnSent : Maybe String
+    }
+
+
+type alias ToastMsg =
+    { onFailure : String
+    , onSuccess : Maybe String
+    }
+
+
+base : Action
+base =
+    { run = \store -> ( store, Cmd.none )
+    , toastOnSent = Nothing
+    }
+
+
+getPosts : Action
+getPosts =
+    { base
+        | run =
+            \store ->
+                if shouldSendRequest store.posts then
+                    ( { store | posts = Loading }
+                    , send "Failed to get posts"
+                        (\err s -> { s | posts = Failure err })
+                        API.Post.getAll
+                        GotPosts
+                    )
+
+                else
+                    ( store, Cmd.none )
```

Comment:
Any thoughts on iterating on this or merging this PR in general?
One of the ways I was thinking about taking this idea even further is moving to a `Reader` pattern or something like `Effect` to pull out the transition logic over the spine of the `Async` type.

You can see how I achieved that with [ReasonMl's GADT syntax here](https://gist.github.com/erlandsona/1fb108d32ca6bf01e49da9521aedc63f#file-reducert-re-L18)
and the subsequent [`epic` that filtered for "Requested" msg's here](https://gist.github.com/erlandsona/1fb108d32ca6bf01e49da9521aedc63f#file-request-re-L13) to modify the spine of the `Async` type any time someone dispatches a request.

One downside to this approach is the subsequent `Parent << Child << SubChildMsg` msg's needed to thread the responses back into the nested stores we were using...

That said, given the Store pattern I think we'd mostly eliminate that because the Store is all top level... only thing we'd need is access to the right `key` in the store to transform.

Of course my answer to that is `elm-accessors` but a tuple of `get : Store -> SubChild.Model` & `over : (SubChild.Model -> SubChild.Model) -> Store -> Store` would behave the same, save for being dramatically more boilerplate than just a simple `Lens.subChild` 🤔